### PR TITLE
Responsive mixin

### PR DESCRIPTION
**Pull request by skitson**
## Description
This PR adds a new responsive mixin that can be used to generate breakpoint specific classes and a general use class. This has been used for the text alignment and hide classes.

Once the responsive mixin has been merged and signed off, we can then go back through the code base and use it to generate numerous responsive classes, and also reduce lines of code in cases such as the responsive width mixin.

The responsive mixin is very flexible because it allows content to be passed into it using `@content`.

## Related issue
https://github.com/sky-uk/toolkit/issues/181

## Motivation and Context
This change was required to allow the use of classes to target specific breakpoints for specific styles. It will help reduce the amount of code written across the estate, provide some flexibility to make any class responsive, and ensure uniformity.

This change supports the Toolkit evolution programme.

## How Has This Been Tested?
This has been tested using Toolbelt applying the generated classes for alignment and hiding to the Hello world component.

Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc.

## Types of Changes
What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Browser Support
- [x] Chrome
- [ ] Edge
- [ ] Firefox
- [ ] IE9
- [ ] IE10
- [ ] IE11
- [ ] Opera
- [ ] Safari
- [ ] Mobile Devices
- [ ] Screen-readers (e.g. JAWS, Apple VoiceOver)

## Checklist
- [x] All code includes full inline documentation (as per the [template](https://github.com/sky-uk/toolkit-core/blob/master/_template.scss)).
- [x] All code conforms to the Toolkit [coding style](https://github.com/sky-uk/toolkit/wiki/Coding-Style).
- [x] All code conforms to [WCAG 2.0 level AA Accessibility Guidelines](https://www.w3.org/TR/WCAG20/).
- [x] New functions and mixins have appropriate tests.
- [x] All new and existing tests passed.
- [ ] I have added instructions on how to test my changes.
- [x] Package version updated.
- [x] CHANGELOG.md updated.

### CODE REVIEW

File: tools/_mixins.scss [hunk 1]
```diff
@@ -92,3 +92,59 @@
...
+//
+// And for each breakpoint in $mq-breakpoints (example below for mobile breakpoint):
+//
+// @media (min-width: 20em) {
+//   .foo\@mobile {
+//     color: red;
+//   }
+// }
+
+@mixin responsive-utility() {
```
Comment (joebell93) [inline tools/_mixins.scss]: 😍

File: tools/_mixins.scss [hunk 2]
```diff
@@ -92,3 +92,59 @@
...
+  }
+}
+
+//  Text alignment mixin to generate breakpoint specific (using the responsive-utility mixin) and general use text alignment classes
+// ===========================================
+
+// To be used in the typography utilities file as so:
+// @include test-alignments(left center right justify);
+
+@mixin text-alignments($alignment-positions) {
```
Comment (joebell93) [inline tools/_mixins.scss]: I think this mixin should probably live within the `_typography.scss` file

We night not even need to necessarily make it a mixin as it isn't too re-usable just yet

File: tools/_mixins.scss (hunk 1 above)
Comment (joebell93) [inline tools/_mixins.scss]: What about `responsive-suite` as a name?
Just a thought

File: tools/_mixins.scss [hunk 3]
```diff
@@ -92,3 +92,59 @@
...
+//   color: red;
+// }
+//
+// And for each breakpoint in $mq-breakpoints (example below for mobile breakpoint):
+//
+// @media (min-width: 20em) {
+//   .foo\@mobile {
+//     color: red;
+//   }
+// }
```
Comment (joebell93) [inline tools/_mixins.scss]: Ace documentation

**Comment by aaronthomas** [comment]
Nice work Sam, I really like this.

One potential concern – could generating a class for _every_ breakpoint be a little excessive and add bloat? For all the ad-hoc responsive classes we've had to create in Mobile Sales, they tend to be for just one particular breakpoint and not needed for the others. Is there a way the user could define the breakpoints they wish to generate classes for? 

This is more of an issue on projects where `$mq-breakpoints` has been extended. For us, we have the four default toolkit, plus another 3-4 project specific ones – I can't image a scenario where we would want a utility for 8 different breakpoints.

**Comment by joebell93** [comment]
@aaronthomas agree with your concerns, what about if you could pass in specific breakpoints to generate?

`all` could generate breakpoint classes for all of Toolkit's `mq-breakpoints`

**Comment by MrDinsdale** [comment]
Taking an array of breakpoints as an option would be nice. Perhaps it could just default to all if non are set.

**Comment by joebell93** [comment]
@MrDinsdale I was thinking the same, but could `all` by default potentially cause more unnecessary bloat? (if bloat is a concern)

**Comment by MrDinsdale** [comment]
Could be something simple like:

``` scss
@mixin responsive-utility($breakpoints: $mq-breakpoints) {
  @at-root #{&} {
    @content;
  }

  @each $alias, $bp in $breakpoints {
    @include mq($from: $bp) {
      @at-root #{&}\@#{$alias} {
        @content;
      }
    }
  }
}```

**Comment by skitson** [comment]
@aaronthomas good point. @MrDinsdale that's a nice solution. I'll take a look at implementing that!

File: tools/_mixins.scss (hunk 3 above)
Comment (skitson) [inline tools/_mixins.scss]: Thanks!

File: tools/_mixins.scss (hunk 2 above)
Comment (skitson) [inline tools/_mixins.scss]: I'll move it across. I think I just like creating mixins but will look at unwrapping it.

**Comment by MrDinsdale** [comment]
I think the trick will be to enable specifying breakpoints but still restricting to the `$mq-breakpoints` map.

File: tools/_mixins.scss (hunk 1 above)
Comment (skitson) [inline tools/_mixins.scss]: If it was entirely up to me I'd call it something 'crazy' and 'wacky' like `responsivizer`.

**Comment by skitson** [comment]
@MrDinsdale yes - that's going to be the best way. I'll get my thinking cap on.

File: tools/_mixins.scss (hunk 1 above)
Comment (joebell93) [inline tools/_mixins.scss]: `responsivize()` sounds lit 🔥

**Comment by joebell93** [comment]
Could we also add it to [`_hide.scss`](https://github.com/sky-uk/toolkit-core/blob/master/utilities/_hide.scss)?

File: tools/_mixins.scss [hunk 4]
```diff
@@ -92,3 +92,99 @@
     clear: both;
   }
 }
+
+// General purpose responsive mixin
```
Comment (joebell93) [inline tools/_mixins.scss]: I reckon just call this `Responsivise`

Also do we fancy being American or are we going to be stubborn brits and use `s` over `z`?

Weirdly feel that `z` is more readable

**Comment by skitson** [comment]
@joebell93 yeah I don't see why not! I'll add it to the `_hide.scss` file too.

File: tools/_mixins.scss (hunk 4 above)
Comment (skitson) [inline tools/_mixins.scss]: Yeah I swapped the z for an s. Felt awkward with the z in there.

File: tools/_mixins.scss [hunk 5]
```diff
@@ -92,3 +92,99 @@
...
+// @media (min-width: 740px) {
+//   .foo\@medium {
+//     color: red;
+//   }
+// }
+//
+// If an argument list is provided but none of the items match any of the keys in the $mq-breakpoints map,
+// the class will be printed once with no media query wrapper:
+//
+// .foo {
```
Comment (joebell93) [inline tools/_mixins.scss]: As much as I enjoyed `nonsense gobbledigook`, I think we could probably trim the inline docs for this bit and cut the example

File: tools/_mixins.scss (hunk 4 above)
Comment (joebell93) [inline tools/_mixins.scss]: @sky-uk/toolkit what do you guys thing?

Personally I feel like `z` is more readable but I'm open to debate 🇺🇸

Comment (lbarratt) [inline tools/_mixins.scss]: Maybe go with `z` since CSS is already americanised?

Comment (joebell93) [inline tools/_mixins.scss]: I mean we are using `color()`, makes sense to keep it consistent

Comment (skitson) [inline tools/_mixins.scss]: Okay, z it is!

File: tools/_mixins.scss (hunk 5 above)
Comment (skitson) [inline tools/_mixins.scss]: Will do

**Comment by skitson** [comment]
@joebell93 yeah - I expected the `@assert` part of the test to print out. I'll get that sorted.

**Comment by lbarratt** [comment]
Tests look pretty sensible and seem to cover all obvious use cases, nice one 👍

**Pull request merged**